### ISSUE TITLE
Mapping Out the Machine

### ISSUE THREAD

**Issue by zoomclub** [nickfargo/state#11]
The structure I have for the FSM is all contained in one object called "machine". This would be the owner and it in turn has eight states that are objects as well. Then inside these eight objects are one more level of objects and then there are the methods. So at the forth level we hit the pay dirt and access functionality. There are a lot of tools in my app so this is the arrangement they form.

Would each level of objects be another state like so?

``` js
state( machine, {
    level1object : state(objectName {
        level2object : state(objectName{
               Formal: {
                   greet: function () { return "How do you do?"; }
               },
              Casual: {
                  greet: function () { return "Hi!"; }
              }
        }),
    }),
    level1object : state...
    ...
});
```

Presently, State.js running under Node.js on my local machine with the basic step 1, 2, 3 intro running fine. I'm architecting an advanced app and have been readying all the possible states and now it is time to create a working state tree. It would be great to be in touch over email where I can share more. You could contact me by typing  "musicinourblood" into Google, accessing the link at the top of the list then sending a friendly message to the dude named Thomas using the contact form :-)

**Comment by nickfargo** [comment]
If I can restate your description and intentions properly, the “levels of objects” you have there are defining nested substates, inside of which you will define methods.

The effect of this, then, will be that your owner `machine` will bear one of these `State`s at a time as its current state, causing any method defined or inherited by that `State` to “appear” as a method of `machine`. When you transition `machine` to a different `State`, then that `State`’s methods will effectively become implemented on `machine`.

This should then take a form like:

``` js
var machine = {}; // or any already-existing object

state( machine, {

  nameOfState: state({

    nameOfSubstate: state({
      method: function () { return "machine acting this way"; }
    }),

    nameOfAnotherSubstate: state({
      method: function () { return "machine acting that way"; }
    })
  }),

  nameOfAnotherState: state({
    // ...
  })
});

machine.state('-> nameOfSubstate');
machine.method();
// >>> "machine acting this way"

machine.state('-> nameOfAnotherSubstate');
machine.method();
// >>> "machine acting that way"
```

Point of syntax: the _name_ of a state is interpreted according to the _key_ to which the state expression _value_ is mapped; the leading string argument is used to designate _attributes_:

``` js
state( owner, {
  name: state( 'initial', {
    // state expression body
  })
});
```

---

If you have stuff you need to keep private, do feel free to just send directly to (my-github-name at gmail). Anything else I’d rather keep here in the open if that’s alright.

**Comment by zoomclub** [comment]
Sounds good!

Since the functionality in my states relates to discrete areas of the app plus might get rather large in code size I was thinking that the 2nd level state under machine would be separate machines on their own, but still managed under the main machine. So there would be eight machines under the roof of the one main machine?

Or another way might be to just have the main machine object be the state and substate tree for all objects and the the method calls would just be refs to methods that are in other object libs? I'm just looking at ways to structure and manage all the states.

Everything in the app is event driven where incoming events from mouse, qwerty, stylus, etc. are parsed and tracked in a tracker object. Specific derived properties from events would then be responsible for transitioning the machine or submachines plus the current state method would then look to the tracker for other live event feeds it requires to do its work with.

My concern is that all states are under one umbrella so that transitioning between the larger state objects and then their substates is seamless. The other architecture issue I face is where to place all the possible methods a state and its substates can have. This is why I'm hoping to go with eight submachine objects with maybe one main machine tree that then refs methods that are contained in object libs outside the main machine, which would then just concern itself with the main state flow?

For remembering where a given state left off and returning to that point I was thinking that I would just store this in the tracker too. I'm not sure if State.js states can record snapshots and reinstate them?

**Comment by nickfargo** [comment]
Again I’ll try to restate to see how much I understand correctly:

It sounds like you have eight separate aspects of your app, each of which could be respresented as a “machine” object. Each machine _M_ must independently react to some incoming event stream, where a response may include changing the state of _M_, causing _M_ to alter how it responds (as expressed by calls to its methods) to various other event streams.

If that much is correct, then some begged questions would be:
- What is the nature of each _M_; can each _M_ extend from a common prototype?
- What are the states of each _M_?
- Are there any states that are common to each _M_?

I think I may be able to answer the rest better once those questions are settled …

**Comment by zoomclub** [comment]
Right now, the eight objects represent collections of functionality that are aligned to UI areas. Some areas are like toolbars and palettes, while others are Z layers and represent content editors. Each M is a form of encapsulating the specific interactive aspects of the app and a doorway to engaging a part of the app. Any state in a M might bridge out to other behind the scene function libraries, but those library objects are not an integral part of a M.

What I had in mind was that each M should work like a larger state, where it can be transitioned to when its corresponding UI area becomes focused (the State.js transition methods could then be used to erect UI areas as well). For the most part each M would have the focus and just perform one state at a time.

There are also some M's that should work separately or concurrently, where each is doing its own transitions, which are triggered and fed by each M's own event stream. An example being one M that the users right hand engages while the other M is engaged by the users left hand to perform calibrations while the right hand carries on painting.

There can be some common prototype on some M's and I can discover that next. Essentially, there are a lot of tool requirements that I'm forming into collections, which become the states. Likely there can be some states in common between M's. I'll bear these guiding principles in mind as I continue to forge forward...

**Comment by nickfargo** [comment]
I think I see. (Do correct me if any of the following is missing the point.)

From this description, it sounds like you may be touching on the two major remaining milestones of the project: [Concurrency](http://statejs.org/docs/#about--roadmap--proposed-features--concurrency) and [History](http://statejs.org/docs/#about--roadmap--proposed-features--history). More on those in a bit.

> What I had in mind was that each M should work like a larger state, where it can be transitioned to when its corresponding UI area becomes focused.

There are probably several ways you could go here. One possibility might be something like this:

``` js
var bigMachine = {};

state( bigMachine, 'abstract', {
  uiAreaOne: state( 'abstract', {
    enter: function () { console.log("uiAreaOne gained focus"); },
    exit: function () { console.log("uiAreaOne lost focus"); },

    A: state( 'default', {
      doSomething: function () {}
    }),
    B: state({
      doSomething: function () {}
    })
  }),

  uiAreaTwo: state( 'abstract', {
    enter: function () { console.log("uiAreaTwo gained focus"); },
    exit: function () { console.log("uiAreaTwo lost focus"); },

    X: state( 'default', {
      doSomething: function () {}
    }),
    Y: state({
      doSomething: function () {}
    })
  })
});
```

This allows for only one of the four concrete states (`A`, `B`, `X`, `Y`) to be current at a time; if either `A` or `B` are current then the `abstract` superstate `uiAreaOne` is _active_, and likewise for `X`, `Y` and `uiAreaTwo`. In this way you could model the notion of “focus” in one particular “area” at a time while also further describing the internal state space of each “area”.

As of now, one shortcoming with this approach is that there is yet no intrinsic ability to _retain_ the currency within one of the abstract superstates (“areas”) after it has “lost focus” and then later “regained focus”. E.g., with state `B` current:

``` js
bigMachine.state('-> B');

bigMachine.state('-> uiAreaTwo');
bigMachine.state();                 // >>> State 'X'

bigMachine.state('-> uiAreaOne');
bigMachine.state();                 // >>> State 'A'
```

whereas you might want `uiAreaOne` to have _retained_ its “internal currency” of state `B`, rather than always default to state `A`.

And here’s where a proper [History](http://statejs.org/docs/#about--roadmap--proposed-features--history) implementation would have come in handy — it would allow you to retain the internal currency. A full treatment of History is non-trivial (and why it’s still a milestone), however, it may be worth investigating whether a satisfactory partial implementation could be cobbled together in the meantime, which would at least get us the “retained internal state” feature.

But for now, a hacky workaround might be possible:

``` js
state( bigMachine, 'abstract', {
  uiAreaOne: ( function () {
    var internalState;
    return state( 'abstract', {
      // Redirect an incoming transition to the retained internal state
      enter: function ( transition ) { transition.go( internalState || 'A' ); },

      // Remember which internal state was current and retain that reference
      exit: function ( transition ) { internalState = transition.origin; },

      A: state({}),  // ...
      B: state({})   // ...
    });
  }() ),
  uiAreaTwo: state() // ...
});


bigMachine.state('-> B');

bigMachine.state('-> uiAreaTwo');
bigMachine.state();                 // >>> State 'X'

bigMachine.state('-> uiAreaOne');
bigMachine.state();                 // >>> State 'B'
```

(It should be noted that, while an approach like this should work for single objects, it would break heritability if implementing state on a prototype.)

On the other hand, it may be preferable to separate the “UI areas” into separate objects, each with their own `state` implementations. This would mean you’d need a “manager” object to designate one of the “areas” as being “focused”, which would require some extra plumbing — but each “area” would remain active as far as State.js is concerned, even while deemed not “focused” by the “manager”.

Which may lead into the other case:

> There are also some M's that should work separately or concurrently, where each is doing its own transitions, which are triggered and fed by each M's own event stream.

Here is where proper [Concurrency](http://statejs.org/docs/#about--roadmap--proposed-features--concurrency) would fit in, which would implement the concepts of “orthogonal regions” into a single `state` system. This, again, could be workaround-ed with separate “little machines” and some aspect-oriented plumbing:

``` js
var bossMachine = {};
state( bossMachine, 'abstract', {
  myWay: state( 'default', {
    aMethod: function ( arg ) { return "My way"; }
  }),
  openToAdvice: state({
    aMethod: function ( arg ) {
      var adviceFromPaintingAspect = littlePaintingMachine.aMethod( arg );
      var adviceFromCalibrationAspect = littleCalibrationMachine.aMethod( arg );

      // weave the bits of advice together in some way
      return adviceFromPaintingAspect + "; " + adviceFromCalibrationAspect;
    }
  })
});

var littlePaintingMachine = {};
state( littlePaintingMachine, 'abstract', {
  C: state( 'default', {
    aMethod: function ( arg ) { return "This way"; }
  }),
  D: state({
    aMethod: function ( arg ) { return "That way"; }
  })
});

var littleCalibrationMachine = {};
state( littleCalibrationMachine, 'abstract', {
  E: state( 'default', {
    aMethod: function ( arg ) { return "Comme ci"; }
  }),
  F: state({
    aMethod: function ( arg ) { return "Comme ça"; }
  })
});

bossMachine.aMethod();                  // >>> "My way"

bossMachine.state('-> openToAdvice');
bossMachine.aMethod();                  // >>> "This way; Comme ci"

littleCalibrationMachine.state('-> F');

bossMachine.aMethod();                  // >>> "This way; Comme ça"
```

This could allow each “little machine” to do its own thing, transitioning to whatever states you define for them, while the App would interact with them, only as necessary, through some “bigger” `bossMachine` or something.

So if any of that rambling example was clear — this approach is one way a person might implement communication between separate/concurrent machines, at least until it gets implemented into the **State** library proper.

**Comment by zoomclub** [comment]
I agree that it is "preferable to separate the “UI areas” into separate objects, each with their own state implementations". If this is done would there still be a issue with concurrency, could each M not just be transitioned and fed by its own event stream? My interact module has a manager and also a tracker object, so there is also a logical place for focusing the different uiArea M objects.

The enter and exit functions to focus a uiArea should work well and same with employing internalState vars. I should mention that at this junction I'm not moving in the direction of implementing states based on prototype. For the most part each uiArea based M does its own unique thing.

Spent some time identifying the M objects that would be able to work together in a concurrent way. This is important for my very hands on app. Each concurrently engaged M can also store its state in the interact module tracker object.

Next up, I'll cleanup my uiArea function sections more and send you a better outline of each M and its states, as well as info on the purpose of the different objects in the interact module...

**Comment by nickfargo** [comment]
Right, I think separate “machine” objects that talk to each other (like the `bossMachine` and `littleMachine`s above) can work fine.

**Comment by zoomclub** [comment]
Our discussion so far has helped me to think further about concurrency and subsequently come up with a better composition for the machines in the interact module, this is summarized below:

The interaction stack is defined by the following objects:

```
• manager     - contains and manages the objects that make up the interaction stack
• provider    - gets pointer, qwerty input, requires optional controllers, registers them on body element
• tracker     - parses raw events, detects UI intersections, stores derived values, triggers states
• machine     - represents three layered machines, active states watch tracker for relevant values
• xyzLibs     - library objects defining functions that state methods reference and pass args to
```

The concurrency is expressed in three machine objects:

```
• M1 - in background, for running automatic tracker invoked tasks and also post edit render related tasks
• M2 - in middleground, for running tool tasks, these modal states correspond to specific UI areas
• M3 - in foreground, for running tool calibration tasks, these complement whatever state is active in M2
```

Due to the event streams constantly being delivered to the tracker via the provider the interaction stack already has a kind of concurrency taking place. After processing raw events the tracker derives more meaningful values and acts like composite event record, a live snapshot of what is happening in the system at each time slice. The tracker is ultimately responsible for triggering all three machines, based on auto determined values or discrete user calibrations. The states of M1 and M3 can be seen as asides (side-effects) that occur in addition or on top of the current state in M2.

Through the interaction stack the provider and tracker are already doing their own thing processing event streams, even while M2 is transitioned into a given state and handling its own methods. This can result in the tracker firing off states in M1 and M2 while also providing live values to the active M2 state. It is useful to know that there are dynamic events cooking and firing in the tracker, but having three concurrent machines is still required to make the apps interactive surface work as intended.

Earlier in time, using Java I prototyped what amounted to a FSM and it was all squeezed into one event driven flow with many asides. Comparativily, the JS language has brought many joys including State.js. Having concurrent machines seems reasonable, they would work like our two hands do, independently and complementary at the same time, plus there would also be M1, which is like having a third invisible hand that automatically runs through its states at tracker determined times in response to system or edit events.

Having three machines with their states and substates each is better than eight. As long as the methods in the substates are able to subscribe to selective dynamic values in the tracker, which are required for their operation, then the state can work as intended. It would be good for states to also recall their exit values (history) for restoration on re-entry but for the most part states are always responding to fresh values in the tracker.

Hope this clarifies things for you as it has for me :-)

**Comment by nickfargo** [comment]
Glad you’ve made progress. I hope you’re able to push this to github at some point; I’d be very interested to see how the implementation goes.

**Comment by zoomclub** [comment]
When do you plan on implementing concurrent states in State.js, so that the three machines in my app can work well together? For the immediate future I am continuing to engineer out the rest of my app with a goal of starting final development very soon. I'm in the process of finding the right tools and API's to help build the best product possible.

**Comment by nickfargo** [comment]
I can say that the planned full treatment of [Concurrency as it is outlined here](http://statejs.org/docs/#about--roadmap--proposed-features--concurrency) is the next major milestone, now that the v0.2 object model is complete — however, when on the calendar that will land would be dangerous to predict, I’m afraid, as there’s really an awful lot of minute details to consider carefully as that goes forward.

Based on what I’ve worked out so far, though, it should be possible to manually implement the conceptual equivalent of state concurrency / orthogonal regions in terms of the machine–submachine pattern described above (`bossMachine` / `littleMachine`). Lost would be the single-tree heritability of `State`s across region–subregion boundaries, and thus the likely desirable features of automatic method dispatch and event propagation that would be baked into “proper” Concurrency; however, these bits could be rigged up manually without too much pain by delegating method calls and events through to submachines, along the lines of the “aspect-weaving” part of the example above (in `bossMachine.state('openToAdvice')`).

And for what it’s worth, when proper Concurrency does land, any such existing implementation should continue to work just fine, and, if it follows that general pattern, should be fairly painless to reduce into terms of `StateExpression`s.

**Comment by zoomclub** [comment]
Sounds good and your proposed solution will work fine until fully featured concurrency is ready. Here I'm going to keep working out the details of the states, substates and their methods and all the workings of the interact module. Then bring that together with the networking and rendering sides of the app. I'm almost done this phase, just a few more weeks to go. I have been doing a lot of exploring and architecting and am now nearing the end of engineering that out, I'll keep you posted on the progress. Cheers!

**Comment by nickfargo** [comment]
Great! Looking forward to it. I’m sure your experience there will be very informative to the eventual Concurrency implementation.

Also perhaps worth a mention:

> [comment](#issuecomment-26440373) — For remembering where a given state left off and returning to that point I was thinking that I would just store this in the tracker too. I'm not sure if State.js states can record snapshots and reinstate them?

This may have been adequately addressed by PR #13 referenced above. I want to give this new `retained` attribute some more scrutiny but I think that should make it into a 0.2.1 release pretty soon.
